src/main/java/frc/robot/subsystems/elevator/Elevator.java

```diff
@@ -0,0 +1,191 @@
+          ElevatorCal.MAX_ACCELERATION_IN_PER_SECOND_SQUARED));
+
+  private ProfiledPIDController currentPIDController = noteScoringElevatorController;
+
+  private TreeMap<ElevatorPosition, Double> elevatorPositions;
+
+  private ElevatorPosition desiredPosition = ElevatorPosition.HOME;
+
+  private boolean usingNoteScoringPID = true;
+
+  private double prevVelocity = 0;
+  private double prevTimestamp = -1;
+
+  public Elevator() {
+    SparkMaxUtils.initWithRetry(this::initSparks, ElevatorCal.MAX_INIT_RETRY_ATTEMPTS);
+    elevatorPositions = new TreeMap<ElevatorPosition, Double>();
+    elevatorPositions.put(ElevatorPosition.HOME, ElevatorCal.POSITION_IN_HOME);
+    elevatorPositions.put(ElevatorPosition.SCORE_AMP, ElevatorCal.POSITION_IN_SCORE_AMP);
+    elevatorPositions.put(ElevatorPosition.SCORE_TRAP, ElevatorCal.POSITION_IN_SCORE_TRAP);
+    elevatorPositions.put(ElevatorPosition.PRE_CLIMB, ElevatorCal.POSITION_IN_PRE_CLIMB);
+    elevatorPositions.put(ElevatorPosition.POST_CLIMB, ElevatorCal.POSITION_IN_POST_CLIMB);
+    this.desiredPosition = ElevatorPosition.HOME;
+  }
+
+  public void useScoringPID(boolean useNoteScoringPID){
+    usingNoteScoringPID = useNoteScoringPID;
+    if(useNoteScoringPID){
+      currentPIDController = noteScoringElevatorController;
+    }
+    else{
+      currentPIDController = chainGrabberElevatorController;
+    }
+  }
+
+  private void controlPosition(ElevatorPosition pos){
+    currentPIDController.setGoal(elevatorPositions.get(pos));
+    double elevatorDemandVolts = currentPIDController.calculate(leftMotorEncoder.getPosition());
+    if(usingNoteScoringPID){
+      elevatorDemandVolts +=
+        ElevatorCal.NOTE_SCORING_FF.calculate(currentPIDController.getSetpoint().velocity);
+      if(prevTimestamp != -1){
+        elevatorDemandVolts += ElevatorCal.NOTE_SCORING_FF.calculate(prevVelocity, currentPIDController.getSetpoint().velocity, Timer.getFPGATimestamp()-prevTimestamp);
+      }
+    }
+    else{
+      elevatorDemandVolts +=
+        ElevatorCal.CLIMBING_FF.calculate(currentPIDController.getSetpoint().velocity);
+      if(prevTimestamp != -1){
+        elevatorDemandVolts += ElevatorCal.CLIMBING_FF.calculate(prevVelocity, currentPIDController.getSetpoint().velocity, Timer.getFPGATimestamp()-prevTimestamp);
+      }
+    }
+    prevVelocity = currentPIDController.getSetpoint().velocity;
+    prevTimestamp = Timer.getFPGATimestamp();
+    leftMotor.setVoltage(elevatorDemandVolts);
+  }
+
+  public void setDesiredPosition(ElevatorPosition desiredPosition){
+    this.desiredPosition = desiredPosition;
+  }
+
+  public boolean atDesiredPosition(){
+      double desiredPositionIn = elevatorPositions.get(desiredPosition);
+      double currentPositionIn = leftMotorEncoder.getPosition();
+      return Math.abs(desiredPositionIn-currentPositionIn) < ElevatorCal.ELEVATOR_MARGIN_IN;
+  }
+
+  @Override
+  public void periodic(){
+    controlPosition(desiredPosition);
```

Comment:
we prob need a function to set desiredPosition

Reply:
you trolled